datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssigner.java

```diff
@@ -5,30 +5,128 @@
+    PartitionThroughputInfo defaultPartitionInfo = new PartitionThroughputInfo(DEFAULT_KB_RATE, DEFAULT_MESSAGE_RATE, "");
+    newPartitions.forEach((task, partitions) -> {
+      int totalThroughput = partitions.stream().
+          mapToInt(p -> partitionInfoMap.getOrDefault(p, defaultPartitionInfo).getBytesInKBRate()).sum();
+      taskThroughputMap.put(task, totalThroughput);
+    });
+    ArrayList<String> sortedTasks =
+        (ArrayList<String>) newPartitions.keySet().stream().sorted(Comparator.comparing(taskThroughputMap::get)).
+             collect(Collectors.toList());
+
+    ArrayList<String> recognizedPartitions = new ArrayList<>(); // partitions with throughput info
+    ArrayList<String> unrecognizedPartitions = new ArrayList<>(); // partitions without throughput info
+    for (String partition : unassignedPartitions) {
+      if (partitionInfoMap.containsKey(partition)) {
+        recognizedPartitions.add(partition);
+      } else {
+        unrecognizedPartitions.add(partition);
+      }
+    }
+
+    // sort unassigned partitions with throughput info on throughput
+    recognizedPartitions.sort((p1, p2) -> {
+      Integer p1KBRate = partitionInfoMap.get(p1).getBytesInKBRate();
+      Integer p2KBRate = partitionInfoMap.get(p2).getBytesInKBRate();
+      return p1KBRate.compareTo(p2KBRate);
+    });
+
+    // assign partitions with throughput info one by one, by putting the heaviest partition in the lightest task
+    while (recognizedPartitions.size() > 0) {
+       String heaviestPartition = recognizedPartitions.remove(recognizedPartitions.size() - 1);
+       int heaviestPartitionThroughput = partitionInfoMap.getOrDefault(heaviestPartition, defaultPartitionInfo)
+           .getBytesInKBRate();
+       String lightestTask = sortedTasks.remove(0);
+       newPartitions.get(lightestTask).add(heaviestPartition);
+       taskThroughputMap.put(lightestTask, taskThroughputMap.get(lightestTask) + heaviestPartitionThroughput);
+       this.insertTaskIntoSortedList(lightestTask, sortedTasks, taskThroughputMap);
+    }
+
+    // TODO implement a mechanism to prevent tasks from having more than partitionsPerTask partitions
+    // assign unrecognized partitions with round-robin
+    int index = 0;
+    for (String partition : unrecognizedPartitions) {
+      String currentTask = sortedTasks.get(index);
+      newPartitions.get(currentTask).add(partition);
+      index = (index + 1) % sortedTasks.size();
+    }
+
+    // build the new assignment using the new partitions for the affected datastream's tasks
+    Map<String, Set<DatastreamTask>> newAssignments = new HashMap<>();
+    currentAssignment.keySet().forEach(instance -> {
+      Set<DatastreamTask> oldTasks = currentAssignment.get(instance);
+      Set<DatastreamTask> newTasks = oldTasks.stream().map(task -> {
+        if (task.getTaskPrefix().equals(datastreamGroupName)) {
+          return new DatastreamTaskImpl((DatastreamTaskImpl) task, newPartitions.get(task.getId()));
```

Comment:
`_maxPartitionsPerTask` check missing

Reply:
Post assignment validations will go into `LoadBasedPartitionAssignmentStrategy`. Separate PR will be published for that.